test/qa/test_login.py

```diff
@@ -137,3 +138,81 @@ def test_logout_disconnects():
+
+    with lib.Defer(selenium.browser_kill):
+        # Get login link from NordVPN app, trim all spaces & chars after link itself
+        login_link = sh.nordvpn.login(login_flag).split(": ")[1][:login.LOGIN_LINK_LENGTH]
+
+        # Open login link from NordVPN app
+        browser.get(login_link)
+
+        # User credentials, that we will use in order to log in to NordAccount
+        USER_INFO = os.environ.get("DEFAULT_LOGIN_USERNAME") + ":" + os.environ.get("DEFAULT_LOGIN_PASSWORD")
+
+        try:
+            # Username page
+            selenium.browser_element_interact(login.NA_USERNAME_PAGE_TEXTBOX_XPATH, USER_INFO.split(':')[0])
+            selenium.browser_element_interact(login.NA_USERNAME_PAGE_BUTTON_XPATH)
+
+            # Password page
+            selenium.browser_element_interact(login.NA_PASSWORD_PAGE_TEXTBOX_XPATH, USER_INFO.split(':')[1])
+            selenium.browser_element_interact(login.NA_PASSWORD_PAGE_BUTTON_XPATH)
+
+            # Continue to app page
+            selenium.browser_element_interact(login.NA_CONTINUE_PAGE_LINK_BUTTON)
+        except:
+            browser.save_screenshot(login.BROWSER_LOGS_PATH + "Screenshot.png")
+            pytest.fail()
+
+    assert login.LOGOUT_MSG_SUCCESS in sh.nordvpn.logout()
+
+
+@pytest.mark.parametrize("login_flag", login.LOGIN_FLAG)
+@pytest.mark.flaky(reruns=2, reruns_delay=90)
+@timeout_decorator.timeout(60)
+def test_selenium_login_callback(login_flag):
+    selenium = login.SeleniumBrowser()
+    browser = selenium.browser_get()
+
+    with lib.Defer(selenium.browser_kill):
+        # Get login link from NordVPN app, trim all spaces & chars after link itself
+        login_link = sh.nordvpn.login(login_flag).split(": ")[1][:login.LOGIN_LINK_LENGTH]
+
+        # Open login link from NordVPN app
+        browser.get(login_link)
+
+        # User credentials, that we will use in order to log in to NordAccount
+        USER_INFO = os.environ.get("DEFAULT_LOGIN_USERNAME") + ":" + os.environ.get("DEFAULT_LOGIN_PASSWORD")
+
+        try:
+            # Username page
+            selenium.browser_element_interact(login.NA_USERNAME_PAGE_TEXTBOX_XPATH, USER_INFO.split(':')[0])
+            selenium.browser_element_interact(login.NA_USERNAME_PAGE_BUTTON_XPATH)
+
+            # Password page
+            selenium.browser_element_interact(login.NA_PASSWORD_PAGE_TEXTBOX_XPATH, USER_INFO.split(':')[1])
+            selenium.browser_element_interact(login.NA_PASSWORD_PAGE_BUTTON_XPATH)
+
+            # Continue to app page
+            # preferences not set in constructor, so when we click link it does not redirect us to app.
+            callback_link = selenium.browser_element_interact(login.NA_CONTINUE_PAGE_LINK_BUTTON, return_attribute='href')
+        except:
+            browser.save_screenshot(login.BROWSER_LOGS_PATH + "Screenshot.png")
+            pytest.fail()
+
+        assert login.LOGIN_MSG_SUCCESS in sh.nordvpn.login("--callback", callback_link)
```

Comment:
If the message changes in the app then `logout` will not be called, and it will remain logged in. wouldn't be better to execute the functions in a `defer` block?

Reply:
Good idea!